src/Command/RunCommand.php

```diff
@@ -699,4 +667,92 @@ private static function getFormatterName(InputInterface $input): FormatterName
 
         return FormatterName::from($value);
     }
+
+    /**
+     * @return array{string, string|null, bool, string|null}
+     */
+    private static function getSourceFilters(InputInterface $input)
+    {
+        $filter = trim((string) $input->getOption(self::OPTION_FILTER));
+
+        [$gitDiffFilter, $isForGitDiffLines, $gitDiffBase] = self::getGitOptions($input);
+
+        self::assertOnlyOneTypeOfFiltering($filter, $gitDiffFilter);
+
+        return [
+            $filter,
+            $gitDiffFilter,
+            $isForGitDiffLines,
+            $gitDiffBase,
+        ];
+    }
+
+    /**
+     * @return array{string|null, bool, string|null}
+     */
+    private static function getGitOptions(InputInterface $input): array
+    {
+        $gitDiffFilter = $input->getOption(self::OPTION_GIT_DIFF_FILTER);
+        $isForGitDiffLines = (bool) $input->getOption(self::OPTION_GIT_DIFF_LINES);
+        $gitDiffBase = $input->getOption(self::OPTION_GIT_DIFF_BASE);
+
+        self::assertOnlyOneTypeOfGitFiltering($gitDiffFilter, $isForGitDiffLines);
+        self::assertGitBaseHasRequiredFilter($gitDiffFilter, $isForGitDiffLines, $gitDiffBase);
+
+        return [$gitDiffFilter, $isForGitDiffLines, $gitDiffBase];
+    }
+
+    private static function assertOnlyOneTypeOfGitFiltering(
+        ?string $gitDiffFilter,
+        bool $isForGitDiffLines,
+    ): void {
+        if ($isForGitDiffLines
+            && $gitDiffFilter !== Container::DEFAULT_GIT_DIFF_FILTER
+        ) {
+            throw new InvalidArgumentException(
+                sprintf(
+                    'The options "--%s" and "--%s" are mutually exclusive. Please use only one of them.',
+                    self::OPTION_GIT_DIFF_LINES,
+                    self::OPTION_GIT_DIFF_FILTER,
+                ),
+            );
+        }
+    }
+
+    private static function assertGitBaseHasRequiredFilter(
+        ?string $gitDiffFilter,
+        bool $isForGitDiffLines,
+        ?string $gitDiffBase,
+    ): void {
+        if ($gitDiffBase !== Container::DEFAULT_GIT_DIFF_BASE
+            && $gitDiffFilter === Container::DEFAULT_GIT_DIFF_FILTER
+            && $isForGitDiffLines === Container::DEFAULT_GIT_DIFF_LINES
```

Comment:
I guess you can fully test these by calling them with reflection.

Reply:
Probably. I'm not too keen on trying to test this until #2488 at the very least. Until then (and it may still be), any sort of test is bound to be extremely hard to read.

I'm seeking to update this though so I may need to do this, we will see.